### PR TITLE
TAS: Sort domains by the level values, change leaves' ids to hostname

### LINKED ISSUE #3706: TAS: Kueue crashes on admittision when kubernetes.io/hostname is in the lowest level
**What happened**:

Kueue crashes with panic when admitting workload, when kuberntes.io/hostname is in the lowest level.

**What you expected to happen**:

No crash.

**How to reproduce it (as minimally and precisely as possible)**:

This commits reproduces the issue: https://github.com/kubernetes-sigs/kueue/commit/5587f0671b6719e4e015450dc8bd8daa7f3aa892

**Anything else we need to know?**:

This is because the usage copies here is based on only the last value (node name).

### PR DESCRIPTION
**Pull request by PBundyra**
… values

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
- Changes domain's ID to hostname if the lowest level in topology is a node
- Sorts domains by level values instead of ids to account for the change about
- Cleanup integration tests

#### Which issue(s) this PR fixes:

Fixes #3706 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/cache/tas_cache_test.go [hunk 1]
```diff
@@ -1072,3 +1072,67 @@ func TestFindTopologyAssignment(t *testing.T) {
 		})
 	}
 }
+
+func TestLexSortDomains(t *testing.T) {
```
Comment (mimowo) [inline pkg/cache/tas_cache_test.go:1076]: I would prefer to test exposed functions, and I believe this function could be made generic. 
I think a good place is `pkg/util/cmp/cmp.go` and name it as `CompareStringSlices()`. I could also make it generic, but can leave it when needed.

File: pkg/cache/tas_flavor_snapshot.go [hunk 2]
```diff
@@ -414,3 +414,29 @@ func (s *TASFlavorSnapshot) notFitMessage(fitCount, totalCount int32) string {
 	}
 	return fmt.Sprintf("topology %q allows to fit only %v out of %v pod(s)", s.topologyName, fitCount, totalCount)
 }
+
+// lexSortDomains return order of two domains ordered by their levelValues
+func lexSortDomains(a, b *domain) int {
```
Comment (mimowo) [inline pkg/cache/tas_flavor_snapshot.go:419]: let's move this function to a dedicated package (rename and accept string slices as suggested in another comment).

File: test/integration/tas/tas_test.go [hunk 3]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+						}).
+						Ready().
+						Obj(),
+				}
+				for _, node := range nodes {
+					gomega.Expect(k8sClient.Create(ctx, &node)).Should(gomega.Succeed())
+					gomega.Expect(k8sClient.Status().Update(ctx, &node)).Should(gomega.Succeed())
+					gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(&node), &node)).Should(gomega.Succeed())
+
+					ginkgo.By(fmt.Sprintf("node %q, taints : %v", node.Name, node.Spec.Taints))
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:580]: drop

File: test/integration/tas/tas_test.go [hunk 4]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
 			})
 		})
 
+		ginkgo.When("Nodes are created before test with the hostname being the lowest level", func() {
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:524]: Actually, I tested locally that the test passes after https://github.com/kubernetes-sigs/kueue/pull/3694, which is not good. 

The key point is that if we apply the skip form the PR above, then the usage coming from admitted workloads is ignored. So, we can exploit it by creating second workload which would not fit, but fits if the usage is ignored.
One way to do it: don't delete `wl1` and bump the count of pods for `wl2` to 4. After this PR wl2 should be inadmissible, but after the skip usage wl2 would get admitted.

File: test/integration/tas/tas_test.go [hunk 5]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+						Queue(localQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+					gomega.Expect(k8sClient.Create(ctx, wl1)).Should(gomega.Succeed())
+				})
+
+				ginkgo.By("verify the workload is inadmissible", func() {
+					util.ExpectPendingWorkloadsMetric(clusterQueue, 0, 1)
+				})
+			})
+
+			ginkgo.It("should admit workload which fits in a required topology domain", func() {
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:632]:
```suggestion
			ginkgo.It("should admit workload which fits", func() {
```

File: test/integration/tas/tas_test.go [hunk 6]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+				var wl1, wl2 *kueue.Workload
+				ginkgo.By("creating a workload which requires block and can fit", func() {
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:634]:
```suggestion
				ginkgo.By("creating a workload which can fit", func() {
```

File: test/integration/tas/tas_test.go [hunk 7]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+						Queue(localQueue.Name).Request(corev1.ResourceCPU, "1").Obj()
+					gomega.Expect(k8sClient.Create(ctx, wl1)).Should(gomega.Succeed())
+				})
+
+				ginkgo.By("verify the workload is admitted", func() {
+					util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, wl1)
+					util.ExpectReservingActiveWorkloadsMetric(clusterQueue, 1)
+				})
+
+				ginkgo.By("delete the admitted workload", func() {
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:652]: Drop it to strenghten the test

File: test/integration/tas/tas_test.go [hunk 8]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+					gomega.Expect(util.DeleteWorkloadsInNamespace(ctx, k8sClient, ns)).Should(gomega.Succeed())
+					util.ExpectReservingActiveWorkloadsMetric(clusterQueue, 0)
+				})
+
+				ginkgo.By("creating second a workload which requires block and can fit", func() {
+					wl2 = testing.MakeWorkload("wl1", ns.Name).
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:658]:
```suggestion
					wl2 = testing.MakeWorkload("wl2", ns.Name).
```

File: test/integration/tas/tas_test.go [hunk 9]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+						PodSets(*testing.MakePodSet("worker-2", 1).
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:659]:
```suggestion
						PodSets(*testing.MakePodSet("worker-2", 4).
```

File: test/integration/tas/tas_test.go [hunk 10]
```diff
@@ -567,6 +521,159 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+							Toleration(corev1.Toleration{
+								Key:      "node.kubernetes.io/not-ready",
+								Operator: corev1.TolerationOpExists,
+							}).
+							Obj()).
+						Queue(localQueue.Name).Request(corev1.ResourceCPU, "1").Obj()
+					gomega.Expect(k8sClient.Create(ctx, wl2)).Should(gomega.Succeed())
+				})
+
+				ginkgo.By("verify the workload is admitted", func() {
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:670]:
```suggestion
				ginkgo.By("verify the workload wl2 is inadmissible", func() {
```
+ proper check

File: test/integration/tas/tas_test.go [hunk 11]
```diff
@@ -811,46 +891,22 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
-							},
-							Conditions: []corev1.NodeCondition{
-								{
-									Type:   corev1.NodeReady,
-									Status: corev1.ConditionTrue,
-								},
-							},
-						},
-					},
+					*testingnode.MakeNode("cpu-node").
```
Comment (mimowo) [inline test/integration/tas/tas_test.go:894]: Let's make these cleanup changes in a separate PR

**Review by mimowo — COMMENTED** [review]
Overall lgtm, proposing cleanup and to strengthen the test

File: pkg/cache/tas_cache_test.go (hunk 1 above)
Comment (PBundyra) [reply pkg/cache/tas_cache_test.go]: Done

File: pkg/cache/tas_flavor_snapshot.go (hunk 2 above)
Comment (PBundyra) [reply pkg/cache/tas_flavor_snapshot.go]: Done

File: test/integration/tas/tas_test.go (hunk 4 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go:523]: Fixed

File: test/integration/tas/tas_test.go (hunk 5 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go]: Done

File: test/integration/tas/tas_test.go (hunk 6 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go]: Done

File: test/integration/tas/tas_test.go (hunk 7 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go]: Done

File: test/integration/tas/tas_test.go (hunk 8 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go]: Done

File: test/integration/tas/tas_test.go (hunk 9 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go]: Done

File: test/integration/tas/tas_test.go (hunk 10 above)
Comment (PBundyra) [reply test/integration/tas/tas_test.go]: Done

**Pull request merged**